### ISSUE TITLE
Wandb only terminates one process when using DDP.

### ISSUE THREAD

**Issue by jopo666** [wandb/wandb#1579]
**Describe the bug**

Each time I run a sweep (one or multiple runs per agent), some processes are left running (these hog 100% the processor they're assigned to and eat up RAM). The data from GPUs is not cleared after finishing a sweep, leading to  `CUDA out of memory` error.  Afterwards I need to kill all processes individually. After stopping the sweep I get this error.

`/home/jpohjone/miniconda3/envs/models/lib/python3.8/multiprocessing/resource_tracker.py:216: UserWarning: resource_tracker: There appear to be 6 leaked semaphore objects to clean up at shutdown`

There is no problem when running a normal run (ie. copying and running the same command that the sweep uses).

**To Reproduce**

I can create  a small example script later I if seems the bug can't be solved otherwise.

**Expected behavior**

Shut down the script correctly after sweep is done.

**Screenshots**

Snippet from `htop`.

<img width="647" alt="htop" src="https://user-images.githubusercontent.com/49716607/101246013-45d18080-3719-11eb-8da8-e9c8879bc244.png">

**Operating system and other versions**
 - OS: Ubuntu 20.04.1 LT
 - wand: 10.12
 - python: 3.8.5
 - torch: 1.7.0
 - miniconda3

**Additional context**

I use four GPUs with DDP to distribute jobs, but the problem persists even when I only use 1 GPU.

**Comment by tyomhak** [comment]
Hi @jopo666 , thanks for filing this issue! We're looking into it — a reproducible script would be very much appreciated!

**Comment by jopo666** [comment]
Found the/a bug!

When I'm using DistributedDataParallel (DDP) to distribute jobs across 4 GPUs and kill **a sweep** with either `ctrl+C` or within my training script (I've implemented my own early stopping), wandb only terminates the main process initiated by DDP and leaves the other 3 running. I'm guessing this also happens when a sweep is using the `hyperband` method, as terminating a sweep is probably done similarly in this situation.

My additional context part was wrong and the problem does seem to only happen with DDP.

**How to fix:**

Make wandb terminate all started processes, not just the main one, when using DDP.

**Comment by dasturge** [comment]
I also have this issue, I am using nn.DataParallel (only one system), with 4 gpus, also noticed that when I kill a sweep it will not kill the spawned workers and they will simply hang indefinitely while reserving GPU memory.

Environment:
ubuntu 16.04  running tmux
GCC 9.3.0
python: 3.8.10
wandb: '0.10.30'
env:
- conda
- pytorch
- ignite

details:
running a sweep using 4 GPUs with nn.DataParallel and 32 workers for the pytorch dataloader.

I'm running a sweep and I've noticed that if I kill an agent for any reason (CTRL+C x2), the spawned processes won't be cleaned up.  I have to find and kill them manually, then I get:

```
user@server:~/my_project$ sudo kill -9 63537 63541 636
46 63649 63650 63651 63653 63661 63662 63665 63667 63674 63698 63699 63700 63701
 63702 63703 63704 63705 63706
user@server:~/my_project$ /home/user/anaconda3/envs/my_env/lib/python3.8/multiprocessing/resource_tracker.py:216: UserWarning: 
resource_tracker: There appear to be 6 leaked semaphore objects to clean up at s
hutdown
  warnings.warn('resource_tracker: There appear to be %d '
```

it does emit this warning only after I force kill the spawned processes.

**Comment by viniciusdsmello** [comment]
I also have this issue, I'm running with wandb ```wandb, version 0.11.0```. Has anyone figured out how to fix it?

**Comment by shariqfarooq123** [comment]
I'm running into exactly the same issue. Is there any solution to this?

**Comment by kptkin** [comment]
Hi @shariqfarooq123, thanks for reporting the issue!

We are working on an improved version for handling multiprocessing which we call `wandb service`
You can read more about it [here](https://github.com/wandb/client/blob/master/docs/dev/wandb-service-user.md). Would you be interested in trying it out and see if it resolves your issue?

**Comment by shariqfarooq123** [comment]
Calling `wandb.finish()` in every process at the end solved the issue for me. Thanks!

**Comment by kptkin** [comment]
This issue should be solved since 0.13+. Closing as done if you still experience an issue please re-open.
